src/api/__test__/pagopaApi.test.ts

```diff
@@ -0,0 +1,99 @@
+import { Server } from 'http';
+import { Millisecond } from 'italia-ts-commons/lib/units';
+import { PaymentManagerClient } from '../pagopa';
+import { defaultRetryingFetch } from '../../utils/fetch';
+import * as myFetch from '../../utils/fetch';
+
+import pm from './pm';
+
+// Client for the PagoPA PaymentManager
+describe('Weak Test Suite', () => {
+  // eslint-disable-next-line functional/no-let
+  let pmMockServer: Server = {} as Server;
+
+  beforeAll(() => {
+    pmMockServer = pm.listen(5000, '127.0.0.1');
```

Comment:
it looks like these are integration tests (no unit), can we mark them as such (ie. pagopaApi.integration.test.ts) in order to tell them from unit tests?

Reply:
The test against PM test server is an integration test.. I agree. But I used the tests against the local servers only as a way of unit testing without mocking fetches. The approach is similar to the one proposed here https://kentcdodds.com/blog/stop-mocking-fetch . Maybe I should separate the two kind of tests in different files.

Reply:
maybe I'm wrong, but it looks like a different approach: in the article you've linked to the author mocks network responses (unit tests) using msw (that's something you can do with [nock](https://github.com/nock/nock) in NodeJS context) while here we are statring a real http server (which is slower = integration test)

Reply:
Obviously you are right: I'm starting a real http server, not mocking its responses. Nevertheless, in my opinion the approach is _similar_ (has the same advantages, minus the performances, plus the flexibility) to the one proposed in the reference, since handlers in the express application just return a response, as if they were mocked. In my opinion, using a mock or another _double_ for a system, like a stub or a fake object, can't change the nature of the test. So if it's a unit test when mocking server responses, it should remain a unit test when using a fake server which replies with the same responses. Obviously if I'm supposed to adhere to a stricter definition of unit testing I will change the file names as you suggested, or if you think I'd better mock the responses I could change the tests. Anyway thank you for the comment, it   stimulated me to think deeper

Reply:
Apart from worst performance, there are some subtle differences with the two approaches:

1. using your approach you have to configure and start a new http server for every test suite (which may run concurrently). this introduces some maintenance complexity since, while your application grows, you'll need to keep the mock server implementation on par (ie. adding middlewares for different output serialization and so on). this effort is avoidable mocking the responses directly;
1. willing or not, you're testing the implementation of the server as well. for example, your tests may fail because of some express misconfiguration or because the server port is busy. that's the reason they're actually _integration_ tests. it's important (in unit tests) to limit the reason a test may fail, while with your approach this surface is broadening

While I see some downsides, I see no real gain in using a real server for testing. Can you point out some?

Reply:
I understand your points, but the main reasons that pushed me to favour a real server are the following
- Test the case of network failure, or server unavailability, cases that in my opinion are of a great value for the citizen, since she should be warned. Mocking the responses assumes that _there is_ a response.
- Use the same test architecture for unit tests (or something that I call unit tests :-) ) and true integration tests, in which the `express` application served by the true http server can be the same or partially the same of the one used in backend

Reply:
that's right, just make this distinction clear (unit vs integration) and don't mix them since integration tests may be ran against production (so we must have a way to group them)